File: frontend/passes/drop_injector.go
package passes

import (
	"fmt"
	"maps"

	"github.com/mattcarp12/maml/frontend/mir"
	"github.com/mattcarp12/maml/frontend/types"
)

func InjectDrops(g *mir.Graph, locals map[string]types.Type, globalLiveness *LivenessResult) {
	env := buildTypeEnv(g, locals)
	views := buildViewSet(g, locals)

	// Build reverse alias map: src -> []dst (all variables that alias src)
	revAliases := make(map[string][]string)
	for dst, src := range globalLiveness.Aliases {
		revAliases[src] = append(revAliases[src], dst)
	}

	dropCounter := 0
	newDropTemp := func() string {
		dropCounter++
		return fmt.Sprintf("_drop_t%d", dropCounter)
	}

	for _, block := range g.SortedBlocks() {
		stmtLiveness := AnalyzeStatementLiveness(block, globalLiveness.LiveOut[block.ID], globalLiveness.Aliases)
		var newStmts []mir.Instruction

		for i, inst := range block.Statements {
			newStmts = append(newStmts, inst)

			dying := make([]string, 0)
			for v := range stmtLiveness.LiveIn[i] {
				if !stmtLiveness.LiveOut[i][v] && !views[v] {
					// Check if any alias of v is still live; if so, skip dropping v
					if !hasLiveAlias(v, stmtLiveness.LiveOut[i], revAliases) {

						// NEW: Suppress drop if consumed by a CallInst
						consumedByCall := false
						if callInst, ok := inst.(*mir.CallInst); ok {
							for argIdx, argVal := range callInst.Arguments {
								if argReg, isReg := argVal.(*mir.Register); isReg && argReg.Name == v {
									// Ensure we don't go out of bounds if intrinsics lack the array
									if argIdx < len(callInst.ArgConsumed) && callInst.ArgConsumed[argIdx] {
										consumedByCall = true
										break
									}
								}
							}
						}

						if !consumedByCall {
							dying = append(dying, v)
						}
					}
				}
			}

			def := getDef(inst)
			if def != "" && def != "_" && !stmtLiveness.LiveOut[i][def] && !views[def] {
				if !hasLiveAlias(def, stmtLiveness.LiveOut[i], revAliases) {
					dying = append(dying, def)
				}
			}

			for _, v := range dying {
				t := env[v]
				if !needsDrop(t) {
					continue
				}

				if moveInst, ok := inst.(*mir.MoveInst); ok && moveInst.Src == v {
					continue
				}

				if storeInst, ok := inst.(*mir.StoreInst); ok {
					if srcReg, isReg := storeInst.Value.(*mir.Register); isReg && srcReg.Name == v {
						continue
					}
				}

				buildRecursiveDrop(v, t, false, &newStmts, newDropTemp, locals)
			}
		}

		termDying := make([]string, 0)
		for v := range stmtLiveness.TermLiveIn {
			if !stmtLiveness.TermLiveOut[v] && !views[v] {
				if !hasLiveAlias(v, stmtLiveness.TermLiveOut, revAliases) {
					termDying = append(termDying, v)
				}
			}
		}

		for _, v := range termDying {
			t := env[v]
			if !needsDrop(t) {
				continue
			}

			if retTerm, ok := block.Terminator.(*mir.ReturnTerminator); ok && retTerm.Value != nil {
				if reg, isReg := retTerm.Value.(*mir.Register); isReg && reg.Name == v {
					continue
				}
			}

			buildRecursiveDrop(v, t, false, &newStmts, newDropTemp, locals)
		}

		block.Statements = newStmts
	}
}

// hasLiveAlias checks if any alias (direct or transitive) of `v` is present in `liveSet`.
// It uses the reverse alias map to traverse the alias graph.
func hasLiveAlias(v string, liveSet map[string]bool, revAliases map[string][]string) bool {
	visited := make(map[string]bool)
	var dfs func(string) bool
	dfs = func(current string) bool {
		if visited[current] {
			return false
		}
		visited[current] = true
		// Check if any alias of current is in liveSet
		for _, alias := range revAliases[current] {
			if liveSet[alias] {
				return true
			}
			// Recurse to transitive aliases (e.g., a -> b -> c)
			if dfs(alias) {
				return true
			}
		}
		return false
	}
	return dfs(v)
}

func buildRecursiveDrop(
	vName string,
	t types.Type,
	isAddr bool,
	stmts *[]mir.Instruction,
	newTemp func() string,
	locals map[string]types.Type,
) {
	if !needsDrop(t) {
		return
	}

	if st, isStruct := t.(*types.StructType); isStruct {
		for i, field := range st.Fields {
			if needsDrop(field.Type) {
				tmpPtr := newTemp()
				locals[tmpPtr] = types.PtrType{} // Register the new temporary

				*stmts = append(*stmts, &mir.FieldAddrInst{
					Dst:        tmpPtr,
					Object:     &mir.Register{Name: vName, Type: t},
					ObjectType: t,
					FieldName:  field.Name,
					FieldIndex: i,
					FieldType:  field.Type,
				})
				buildRecursiveDrop(tmpPtr, field.Type, true, stmts, newTemp, locals)
			}
		}
		return
	}

	if arr, isArray := t.(*types.ArrayType); isArray {
		for i := 0; i < arr.Size; i++ {
			tmpPtr := newTemp()
			locals[tmpPtr] = types.PtrType{} // Register the new temporary

			*stmts = append(*stmts, &mir.IndexAddrInst{
				Dst:        tmpPtr,
				Source:     &mir.Register{Name: vName, Type: t},
				SourceType: t,
				Index:      &mir.IntConstant{Value: int64(i), Type: types.I64Type{}},
				Type:       arr.Base,
			})
			buildRecursiveDrop(tmpPtr, arr.Base, true, stmts, newTemp, locals)
		}
		return
	}

	symbol := lookupDestructorSymbol(t)
	if symbol != "" {
		var callArgName string

		if isAddr {
			callArgName = vName
		} else {
			// We have the value itself – take its address.
			ptrTmp := newTemp()
			locals[ptrTmp] = types.PtrType{}
			*stmts = append(*stmts, &mir.BorrowInst{
				Dst:   ptrTmp,
				Src:   vName,
				IsMut: false,
			})
			callArgName = ptrTmp
		}

		*stmts = append(*stmts, &mir.CallInst{
			Dst:       "_",
			Function:  &mir.Register{Name: symbol, Type: types.PtrType{}},
			Arguments: []mir.Value{&mir.Register{Name: callArgName, Type: types.PtrType{}}},
			Type:      types.UnitType{},
		})
	}

}

func lookupDestructorSymbol(t types.Type) string {
	switch t.(type) {
	case *types.VectorType:
		return "maml_vec_free"
	case *types.MapType:
		return "maml_map_free"
	case *types.FutureType:
		return "maml_task_release"
	case types.StringType:
		return "maml_str_free"
	default:
		return "maml_free"
	}
}

func buildViewSet(g *mir.Graph, locals map[string]types.Type) map[string]bool {
	views := make(map[string]bool)
	addrIsDerived := make(map[string]bool)

	// 1. Any local variable that is a pointer type is fundamentally a borrow/reference.
	// This naturally covers function parameters since they are mapped into locals.
	for name, t := range locals {
		if _, isPtr := t.(types.PtrType); isPtr {
			addrIsDerived[name] = true
		}
	}

	for _, p := range g.Params {
		if _, isPtr := p.Type.(types.PtrType); isPtr {
			addrIsDerived[p.Name] = true
		}
	}

	for _, block := range g.Blocks {
		for _, inst := range block.Statements {
			switch i := inst.(type) {
			case *mir.FieldAddrInst:
				addrIsDerived[i.Dst] = true
			case *mir.IndexAddrInst:
				addrIsDerived[i.Dst] = true
			case *mir.BorrowInst:
				addrIsDerived[i.Dst] = true
			case *mir.LoadPtrInst:
				// 2. If we load a concrete struct from a borrowed pointer,
				// that local struct is just a view and does NOT own the memory.
				if ptrReg, ok := i.Ptr.(*mir.Register); ok && addrIsDerived[ptrReg.Name] {
					views[i.Dst] = true
				}
			}
		}
	}

	return views
}

func buildTypeEnv(g *mir.Graph, locals map[string]types.Type) map[string]types.Type {
	env := make(map[string]types.Type)
	maps.Copy(env, locals)
	for _, p := range g.Params {
		env[p.Name] = p.Type
	}
	for _, block := range g.Blocks {
		for _, inst := range block.Statements {
			if c, ok := inst.(*mir.CastInst); ok {
				env[c.Dst] = c.Type
			}
			if f, ok := inst.(*mir.FieldAddrInst); ok {
				env[f.Dst] = types.PtrType{}
			}
			if idx, ok := inst.(*mir.IndexAddrInst); ok {
				env[idx.Dst] = types.PtrType{}
			}
			if b, ok := inst.(*mir.BitcastPtrInst); ok {
				env[b.Dst] = b.Type
			}
		}
	}
	return env
}

func needsDrop(t types.Type) bool {
	if t == nil || isPrimitive(t) {
		return false
	}

	switch ty := t.(type) {
	case *types.ViewType:
		return false

	case *types.VectorType, *types.MapType, *types.FutureType, types.StringType:
		return true

	case *types.StructType:
		for _, field := range ty.Fields {
			if needsDrop(field.Type) {
				return true
			}
		}
		return false

	case *types.ArrayType:
		return needsDrop(ty.Base)

	case *types.SumType:
		for _, variant := range ty.Variants {
			for _, field := range variant.Fields {
				if needsDrop(field.Type) {
					return true
				}
			}
			for _, tupleTy := range variant.TupleTypes {
				if needsDrop(tupleTy) {
					return true
				}
			}
		}
		return false

	default:
		return true
	}
}

func isPrimitive(t types.Type) bool {
	switch t.(type) {
	case types.I8Type, types.I16Type, types.I32Type, types.I64Type, types.I128Type,
		types.U8Type, types.U16Type, types.U32Type, types.U64Type, types.U128Type,
		types.F32Type, types.F64Type, types.BoolType, types.CharType, types.UnitType,
		types.PtrType:
		return true
	default:
		return false
	}
}

func getDef(inst mir.Instruction) string {
	switch i := inst.(type) {
	case *mir.AssignInst:
		return i.Dst
	case *mir.CopyInst:
		return i.Dst
	case *mir.MoveInst:
		return i.Dst
	case *mir.BorrowInst:
		return i.Dst
	case *mir.CallInst:
		return i.Dst
	case *mir.FieldAddrInst:
		return i.Dst
	case *mir.BinaryOpInst:
		return i.Dst
	case *mir.UnaryOpInst:
		return i.Dst
	case *mir.CastInst:
		return i.Dst
	case *mir.LoadPtrInst:
		return i.Dst
	case *mir.BitcastPtrInst:
		return i.Dst
	}
	return ""
}

File: passes/src/drop_injector.cpp
#include "drop_injector.h"

namespace maml::passes {

template <class... Ts> struct overloaded : Ts... {
    using Ts::operator()...;
};
template <class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

void DropInjector::injectDrops(mir::Graph* g, std::unordered_map<SymID, const types::Type*>& locals,
    const LivenessResult& globalLiveness)
{
    auto env = buildTypeEnv(g, locals);
    auto views = buildViewSet(g, locals);

    // Build reverse alias map: src -> []dst (all variables that alias src)[cite: 3]
    std::unordered_map<SymID, std::vector<SymID>> revAliases;
    for (const auto& [dst, src] : globalLiveness.aliases) {
        revAliases[src].push_back(dst);
    }

    LivenessAnalyzer livenessAnalyzer(sym_);

    for (mir::BasicBlock* block : g->sortedBlocks()) {
        auto stmtLiveness = livenessAnalyzer.analyzeStatementLiveness(
            block, globalLiveness.liveOut.at(block->id), globalLiveness.aliases);
        std::vector<mir::Instruction> newStmts;

        for (size_t i = 0; i < block->statements.size(); ++i) {
            const auto& inst = block->statements[i];
            newStmts.push_back(inst);
            Position pos = mir::getPosOf(inst);

            std::vector<SymID> dying;
            for (SymID v : stmtLiveness.liveIn[i]) {
                if (stmtLiveness.liveOut[i].find(v) == stmtLiveness.liveOut[i].end()
                    && views.find(v) == views.end()) {
                    if (!hasLiveAlias(v, stmtLiveness.liveOut[i], revAliases)) {

                        // Suppress drop if consumed by a CallInst[cite: 3]
                        bool consumedByCall = false;
                        if (auto* callInst = std::get_if<mir::CallInst>(&inst)) {
                            for (size_t argIdx = 0; argIdx < callInst->arguments.size(); ++argIdx) {
                                if (auto* argReg
                                    = std::get_if<mir::Register>(&callInst->arguments[argIdx])) {
                                    if (argReg->name == v && argIdx < callInst->argConsumed.size()
                                        && callInst->argConsumed[argIdx]) {
                                        consumedByCall = true;
                                        break;
                                    }
                                }
                            }
                        }

                        if (!consumedByCall) {
                            dying.push_back(v);
                        }
                    }
                }
            }

            SymID def = getDef(inst);
            if (def != NoSymbol
                && stmtLiveness.liveOut[i].find(def) == stmtLiveness.liveOut[i].end()
                && views.find(def) == views.end()) {
                if (!hasLiveAlias(def, stmtLiveness.liveOut[i], revAliases)) {
                    dying.push_back(def);
                }
            }

            for (SymID v : dying) {
                auto it = env.find(v);
                if (it == env.end() || !needsDrop(it->second))
                    continue;

                if (auto* moveInst = std::get_if<mir::MoveInst>(&inst)) {
                    if (moveInst->src == v)
                        continue;
                }
                if (auto* storeInst = std::get_if<mir::StoreInst>(&inst)) {
                    if (auto* srcReg = std::get_if<mir::Register>(&storeInst->value)) {
                        if (srcReg->name == v)
                            continue;
                    }
                }

                buildRecursiveDrop(v, it->second, false, newStmts, locals, pos);
            }
        }

        std::vector<SymID> termDying;
        for (SymID v : stmtLiveness.termLiveIn) {
            if (stmtLiveness.termLiveOut.find(v) == stmtLiveness.termLiveOut.end()
                && views.find(v) == views.end()) {
                if (!hasLiveAlias(v, stmtLiveness.termLiveOut, revAliases)) {
                    termDying.push_back(v);
                }
            }
        }

        Position termPos = mir::getPosOf(block->terminator);
        for (SymID v : termDying) {
            auto it = env.find(v);
            if (it == env.end() || !needsDrop(it->second))
                continue;

            if (auto* retTerm = std::get_if<mir::ReturnTerminator>(&block->terminator)) {
                if (auto* reg = std::get_if<mir::Register>(&retTerm->value)) {
                    if (reg->name == v)
                        continue;
                }
            }

            buildRecursiveDrop(v, it->second, false, newStmts, locals, termPos);
        }

        block->statements = std::move(newStmts);
    }
}

bool DropInjector::hasLiveAlias(SymID v, const std::unordered_set<SymID>& liveSet,
    const std::unordered_map<SymID, std::vector<SymID>>& revAliases) const
{
    std::unordered_set<SymID> visited;

    auto dfs = [&](auto& self, SymID current) -> bool {
        if (visited.find(current) != visited.end())
            return false;
        visited.insert(current);

        auto it = revAliases.find(current);
        if (it != revAliases.end()) {
            for (SymID alias : it->second) {
                if (liveSet.find(alias) != liveSet.end())
                    return true;
                if (self(self, alias))
                    return true;
            }
        }
        return false;
    };

    return dfs(dfs, v);
}

SymID DropInjector::newDropTemp()
{
    dropCounter_++;
    return sym_.intern("_drop_t" + std::to_string(dropCounter_));
}

void DropInjector::buildRecursiveDrop(SymID vName, const types::Type* t, bool isAddr,
    std::vector<mir::Instruction>& stmts, std::unordered_map<SymID, const types::Type*>& locals,
    Position pos)
{
    if (!needsDrop(t))
        return;

    if (t->kind == types::TypeKind::Struct) {
        const auto& fields = std::get<types::StructPayload>(t->payload).fields;
        for (size_t i = 0; i < fields.size(); ++i) {
            if (needsDrop(fields[i].type)) {
                SymID tmpPtr = newDropTemp();
                locals[tmpPtr] = reg_.getPrimitive(types::TypeKind::Ptr);

                stmts.push_back(mir::FieldAddrInst { tmpPtr, mir::Register { vName, t, pos }, t,
                    fields[i].name, static_cast<int>(i), fields[i].type, {}, pos });
                buildRecursiveDrop(tmpPtr, fields[i].type, true, stmts, locals, pos);
            }
        }
        return;
    }

    if (t->kind == types::TypeKind::Array) {
        const auto& payload = std::get<types::ArrayPayload>(t->payload);
        for (int i = 0; i < payload.size; i++) {
            SymID tmpPtr = newDropTemp();
            locals[tmpPtr] = reg_.getPrimitive(types::TypeKind::Ptr);

            stmts.push_back(mir::IndexAddrInst { tmpPtr, mir::Register { vName, t, pos }, t,
                mir::IntConstant {
                    static_cast<int64_t>(i), reg_.getPrimitive(types::TypeKind::I64), pos },
                payload.base, pos });
            buildRecursiveDrop(tmpPtr, payload.base, true, stmts, locals, pos);
        }
        return;
    }

    std::string_view symbol = lookupDestructorSymbol(t);
    if (!symbol.empty()) {
        SymID callArgName = NoSymbol;

        if (isAddr) {
            callArgName = vName;
        } else {
            SymID ptrTmp = newDropTemp();
            locals[ptrTmp] = reg_.getPrimitive(types::TypeKind::Ptr);
            stmts.push_back(mir::BorrowInst { ptrTmp, false, vName, pos });
            callArgName = ptrTmp;
        }

        std::vector<mir::Value> args
            = { mir::Register { callArgName, reg_.getPrimitive(types::TypeKind::Ptr), pos } };
        std::vector<bool> consumed = { false };

        stmts.push_back(mir::CallInst { sym_.intern("_"),
            mir::Register {
                sym_.intern(std::string(symbol)), reg_.getPrimitive(types::TypeKind::Ptr), pos },
            args, consumed, reg_.getPrimitive(types::TypeKind::Unit), pos });
    }
}

std::string_view DropInjector::lookupDestructorSymbol(const types::Type* t) const
{
    switch (t->kind) {
    case types::TypeKind::Vector:
        return "maml_vec_free";
    case types::TypeKind::Map:
        return "maml_map_free";
    case types::TypeKind::Future:
        return "maml_task_release";
    case types::TypeKind::String:
        return "maml_str_free";
    case types::TypeKind::Sum:
        return "";
    default:
        return "maml_free";
    }
}

std::unordered_set<SymID> DropInjector::buildViewSet(
    const mir::Graph* g, const std::unordered_map<SymID, const types::Type*>& locals) const
{
    std::unordered_set<SymID> views;
    std::unordered_set<SymID> addrIsDerived;

    for (const auto& [name, t] : locals) {
        if (t && t->kind == types::TypeKind::Ptr)
            addrIsDerived.insert(name);
    }
    for (const auto& p : g->params) {
        if (p.type && p.type->kind == types::TypeKind::Ptr)
            addrIsDerived.insert(p.name);
    }

    for (const auto& [_, block] : g->blocks) {
        for (const auto& inst : block->statements) {
            std::visit(overloaded { [&](std::monostate) {},
                           [&](const mir::FieldAddrInst& i) { addrIsDerived.insert(i.dst); },
                           [&](const mir::IndexAddrInst& i) { addrIsDerived.insert(i.dst); },
                           [&](const mir::BorrowInst& i) { addrIsDerived.insert(i.dst); },
                           [&](const mir::LoadPtrInst& i) {
                               if (auto* ptrReg = std::get_if<mir::Register>(&i.ptr)) {
                                   if (addrIsDerived.find(ptrReg->name) != addrIsDerived.end()) {
                                       views.insert(i.dst);
                                   }
                               }
                           },
                           [&](auto&) {} },
                inst);
        }
    }
    return views;
}

std::unordered_map<SymID, const types::Type*> DropInjector::buildTypeEnv(
    const mir::Graph* g, const std::unordered_map<SymID, const types::Type*>& locals) const
{
    std::unordered_map<SymID, const types::Type*> env = locals;
    for (const auto& p : g->params)
        env[p.name] = p.type;

    for (const auto& [_, block] : g->blocks) {
        for (const auto& inst : block->statements) {
            std::visit(
                overloaded { [&](std::monostate) {},
                    [&](const mir::CastInst& i) { env[i.dst] = i.type; },
                    [&](const mir::FieldAddrInst& i) {
                        env[i.dst] = reg_.getPrimitive(types::TypeKind::Ptr);
                    },
                    [&](const mir::IndexAddrInst& i) {
                        env[i.dst] = reg_.getPrimitive(types::TypeKind::Ptr);
                    },
                    [&](const mir::BitcastPtrInst& i) { env[i.dst] = i.type; }, [&](auto&) {} },
                inst);
        }
    }
    return env;
}

bool DropInjector::needsDrop(const types::Type* t) const
{
    if (!t || isPrimitive(t))
        return false;

    switch (t->kind) {
    case types::TypeKind::View:
        return false;
    case types::TypeKind::Vector:
    case types::TypeKind::Map:
    case types::TypeKind::Future:
    case types::TypeKind::String:
        return true;
    case types::TypeKind::Struct:
        for (const auto& field : std::get<types::StructPayload>(t->payload).fields) {
            if (needsDrop(field.type))
                return true;
        }
        return false;
    case types::TypeKind::Array:
        return needsDrop(std::get<types::ArrayPayload>(t->payload).base);
    case types::TypeKind::Sum:
        for (const auto& variant : std::get<types::SumPayload>(t->payload).variants) {
            for (const auto& field : variant.fields) {
                if (needsDrop(field.type))
                    return true;
            }
            for (const auto& tupleTy : variant.tupleTypes) {
                if (needsDrop(tupleTy))
                    return true;
            }
        }
        return false;
    default:
        return true;
    }
}

bool DropInjector::isPrimitive(const types::Type* t) const
{
    switch (t->kind) {
    case types::TypeKind::I8:
    case types::TypeKind::I16:
    case types::TypeKind::I32:
    case types::TypeKind::I64:
    case types::TypeKind::I128:
    case types::TypeKind::U8:
    case types::TypeKind::U16:
    case types::TypeKind::U32:
    case types::TypeKind::U64:
    case types::TypeKind::U128:
    case types::TypeKind::F32:
    case types::TypeKind::F64:
    case types::TypeKind::Bool:
    case types::TypeKind::Char:
    case types::TypeKind::Unit:
    case types::TypeKind::Ptr:
        return true;
    default:
        return false;
    }
}

SymID DropInjector::getDef(const mir::Instruction& inst) const
{
    return std::visit(overloaded { [&](std::monostate) -> SymID { return NoSymbol; },
                          [&](const mir::AllocaInst& i) { return i.dst; },
                          [&](const mir::AssignInst& i) { return i.dst; },
                          [&](const mir::CopyInst& i) { return i.dst; },
                          [&](const mir::MoveInst& i) { return i.dst; },
                          [&](const mir::BorrowInst& i) { return i.dst; },
                          [&](const mir::CallInst& i) { return i.dst; },
                          [&](const mir::FieldAddrInst& i) { return i.dst; },
                          [&](const mir::BinaryOpInst& i) { return i.dst; },
                          [&](const mir::UnaryOpInst& i) { return i.dst; },
                          [&](const mir::CastInst& i) { return i.dst; },
                          [&](const mir::LoadPtrInst& i) { return i.dst; },
                          [&](const mir::BitcastPtrInst& i) { return i.dst; },
                          [&](const mir::IndexAddrInst& i) { return i.dst; },
                          [&](auto&) -> SymID { return NoSymbol; } },
        inst);
}

} // namespace maml::passes